File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/mq/MessageBuilder.kt
package io.github.kapu.turtlesoup.mq

import io.github.kapu.turtlesoup.config.GameConstants
import io.github.kapu.turtlesoup.models.GameState
import io.github.kapu.turtlesoup.models.HistoryEntry
import io.github.kapu.turtlesoup.utils.MessageKeys
import io.github.kapu.turtlesoup.utils.MessageProvider

/** 메시지 빌드 헬퍼 */
class MessageBuilder(
    private val messageProvider: MessageProvider,
) {
    /** 상태 헤더 빌드 */
    fun buildStatusHeader(state: GameState): String =
        messageProvider.get(
            MessageKeys.ANSWER_HISTORY_HEADER,
            "questionCount" to state.questionCount.toString(),
            "hintCount" to state.hintsUsed.toString(),
            "maxHints" to GameConstants.MAX_HINTS.toString(),
        )

    fun buildAnswerWithHistory(
        state: GameState,
        history: List<HistoryEntry>,
    ): String {
        val header = buildStatusHeader(state)
        val historyLines =
            history
                .mapIndexed { index, item ->
                    messageProvider.get(
                        MessageKeys.ANSWER_HISTORY_ITEM,
                        "number" to (index + 1).toString(),
                        "question" to item.question,
                        "answer" to item.answer,
                    )
                }.joinToString("\n")

        return "$header\n$historyLines"
    }

    fun buildSummary(history: List<HistoryEntry>): String {
        if (history.isEmpty()) {
            return messageProvider.get(MessageKeys.SUMMARY_EMPTY)
        }

        val header = messageProvider.get(MessageKeys.SUMMARY_HEADER, "count" to history.size.toString())
        val body =
            history.mapIndexed { index, item ->
                messageProvider.get(
                    MessageKeys.SUMMARY_ITEM,
                    "number" to (index + 1).toString(),
                    "question" to item.question,
                    "answer" to item.answer,
                )
            }.joinToString("\n")

        return "$header\n$body"
    }

    /** 힌트 블록 빌드 */
    fun buildHintBlock(hints: List<String>): String =
        if (hints.isNotEmpty()) {
            val items =
                hints.mapIndexed { index, hint ->
                    messageProvider.get(
                        MessageKeys.HINT_ITEM,
                        "number" to (index + 1).toString(),
                        "content" to hint,
                    )
                }.joinToString("\n")

            messageProvider.get(
                MessageKeys.HINT_SECTION_USED,
                "hintCount" to hints.size.toString(),
                "hintList" to items,
            )
        } else {
            messageProvider.get(MessageKeys.HINT_SECTION_NONE)
        }
}

File: game-bot-go/internal/turtlesoup/mq/message_builder.go
package mq

import (
	"strings"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/messageprovider"
	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tsmessages "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/messages"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
)

// MessageBuilder 는 타입이다.
type MessageBuilder struct {
	provider *messageprovider.Provider
}

// NewMessageBuilder 는 동작을 수행한다.
func NewMessageBuilder(provider *messageprovider.Provider) *MessageBuilder {
	return &MessageBuilder{provider: provider}
}

// BuildStatusHeader 는 동작을 수행한다.
func (b *MessageBuilder) BuildStatusHeader(state tsmodel.GameState) string {
	return b.provider.Get(
		tsmessages.AnswerHistoryHeader,
		messageprovider.P("questionCount", state.QuestionCount),
		messageprovider.P("hintCount", state.HintsUsed),
		messageprovider.P("maxHints", tsconfig.GameMaxHints),
	)
}

// BuildSummary 는 동작을 수행한다.
func (b *MessageBuilder) BuildSummary(history []tsmodel.HistoryEntry) string {
	if len(history) == 0 {
		return b.provider.Get(tsmessages.SummaryEmpty)
	}

	header := b.provider.Get(tsmessages.SummaryHeader, messageprovider.P("count", len(history)))
	lines := make([]string, 0, len(history))
	for i, item := range history {
		lines = append(lines, b.provider.Get(
			tsmessages.SummaryItem,
			messageprovider.P("number", i+1),
			messageprovider.P("question", item.Question),
			messageprovider.P("answer", item.Answer),
		))
	}
	return header + "\n" + strings.Join(lines, "\n")
}

// BuildHintBlock 는 동작을 수행한다.
func (b *MessageBuilder) BuildHintBlock(hints []string) string {
	if len(hints) == 0 {
		return b.provider.Get(tsmessages.HintSectionNone)
	}

	lines := make([]string, 0, len(hints))
	for i, hint := range hints {
		lines = append(lines, b.provider.Get(
			tsmessages.HintItem,
			messageprovider.P("number", i+1),
			messageprovider.P("content", hint),
		))
	}

	return b.provider.Get(
		tsmessages.HintSectionUsed,
		messageprovider.P("hintCount", len(hints)),
		messageprovider.P("hintList", strings.Join(lines, "\n")),
	)
}
